File: dashboard/lib/media-limits.ts
/**
 * Per-platform, per-surface media limits — the SAME media-limits.json the Python worker
 * reads (worker/media_limits.py). Not a mirror: one file, two readers. That is the whole
 * point — a hand-maintained copy of ~100 numbers drifts, and a drifted number is silent
 * in the worst way (the composer says fine, the worker refuses).
 *
 * ABSENT MEANS NOT ENFORCED. A limit we cannot verify is omitted, never guessed.
 *
 * The file lives at dashboard/media-limits.json — a plain LOCAL import through the
 * existing "@/" alias (tsconfig paths: "@/*" -> "./*", i.e. dashboard/). There is no
 * "@shared/*" alias: Task 1 tried putting this file at the repo root behind one, but
 * Turbopack refuses to bundle anything outside dashboard/ (its project root), and
 * widening turbopack.root pulled the owner's 1.6GB data/ asset store into the build's
 * file-trace manifest. So the file moved INTO dashboard/ instead, and worker/media_limits.py
 * reaches one directory up to read it. See that module's docstring for the full story.
 */
import raw from "@/media-limits.json" with { type: "json" };
import { PLATFORMS, supportsImages, supportsStory, videoSurfaces } from "@/lib/platforms";

export type Violation = {
  kind: "too_short" | "too_long" | "too_small" | "too_large" | "wrong_aspect" | "wrong_format";
  message: string;
  severity: "refuse" | "warn";
};

type Entry = {
  min_duration_ms?: number; max_duration_ms?: number;
  min_width?: number; min_height?: number;
  max_width?: number; max_height?: number;
  min_aspect?: [number, number]; max_aspect?: [number, number];
  max_bytes?: number; formats?: string[];
  note: string; varies?: string;
};

export type AssetLike = {
  media_kind?: string | null;
  duration_ms?: number | null;
  width?: number | null;
  height?: number | null;
  byte_size?: number | null;
};

type RawLimits = {
  schema_version: number;
  platforms: Record<string, Record<string, Record<string, Entry>>>;
};

export type PlatformsData = RawLimits["platforms"];

// Cast through unknown: the imported JSON literal's inferred type (e.g. min_aspect as
// number[]) is narrower in shape than the field-by-field Entry type below, but the
// actual values are validated at load time by worker/media_limits.py's schema check —
// both languages read the identical file.
const REAL_PLATFORMS: PlatformsData = (raw as unknown as RawLimits).platforms ?? {};

// `data` defaults to the real, on-disk media-limits.json for every production call site.
// The parameter exists so anyDestinationAccepts's REFUSAL branch (below) can be proven
// with injected data instead of today's real data — see that function's comment for why
// the real data can never exercise that branch on this install.
export function limitsFor(
  platform: string, surface: string, mediaKind: string, data: PlatformsData = REAL_PLATFORMS,
): Entry | null {
  const p = data?.[platform];
  return p?.[surface]?.[mediaKind] ?? null;
}

/** Cross-multiplied so the comparison is exact. A decimal ratio can exclude 16:9 by
 *  rounding, and 16:9 is a value Meta explicitly permits. Bounds are INCLUSIVE. */
function ratioBelow(w: number, h: number, [aw, ah]: [number, number]): boolean {
  return w * ah < aw * h;
}
function ratioAbove(w: number, h: number, [aw, ah]: [number, number]): boolean {
  return w * ah > aw * h;
}

export function checkMedia(
  platform: string, surface: string, asset: AssetLike, data: PlatformsData = REAL_PLATFORMS,
): Violation[] {
  const entry = limitsFor(platform, surface, asset.media_kind ?? "", data);
  if (!entry) return [];
  // A limit that VARIES by account can never be enforced honestly — warn, never refuse.
  const severity: Violation["severity"] = entry.varies ? "warn" : "refuse";
  const out: Violation[] = [];

  // Every check is guarded on the value being KNOWN. Unknown metadata must never refuse:
  // assets predating the video pipeline carry no duration at all. A `null` must never be
  // coerced to 0 and read as "too short" or "too small" — that is a single `<` away in JS.
  const d = asset.duration_ms;
  if (d != null) {
    if (entry.min_duration_ms != null && d < entry.min_duration_ms) {
      out.push({ kind: "too_short", message: `shorter than ${entry.min_duration_ms / 1000}s`, severity });
    }
    if (entry.max_duration_ms != null && d > entry.max_duration_ms) {
      out.push({ kind: "too_long", message: `longer than ${entry.max_duration_ms / 1000}s`, severity });
    }
  }

  const w = asset.width, h = asset.height;
  if (w != null && h != null) {
    if ((entry.min_width != null && w < entry.min_width) ||
        (entry.min_height != null && h < entry.min_height)) {
      out.push({ kind: "too_small", message: `smaller than ${entry.min_width ?? "?"}x${entry.min_height ?? "?"}`, severity });
    }
    if ((entry.max_width != null && w > entry.max_width) ||
        (entry.max_height != null && h > entry.max_height)) {
      out.push({ kind: "too_large", message: `larger than ${entry.max_width ?? "?"}x${entry.max_height ?? "?"}`, severity });
    }
    if ((entry.min_aspect && ratioBelow(w, h, entry.min_aspect)) ||
        (entry.max_aspect && ratioAbove(w, h, entry.max_aspect))) {
      out.push({ kind: "wrong_aspect", message: `aspect ratio ${w}x${h}`, severity });
    }
  }

  const b = asset.byte_size;
  if (b != null && entry.max_bytes != null && b > entry.max_bytes) {
    out.push({ kind: "too_large", message: `larger than ${entry.max_bytes} bytes`, severity });
  }

  return out;
}

/** The reason this destination cannot take this asset, or null when it can.
 *  Only "refuse" violations disable a chip — a "warn" (a limit that varies by account,
 *  like Discord's) is shown but never blocks the send. Generalizes what used to be
 *  facebook-reel-spec.ts's facebookReelDisabledReason to every platform/surface this
 *  file has an entry for — Facebook Reels is now one case of the general rule, not a
 *  special one. */
export function destinationDisabledReason(
  platform: string, surface: string, asset: AssetLike,
): string | null {
  const refusals = checkMedia(platform, surface, asset).filter((v) => v.severity === "refuse");
  if (refusals.length === 0) return null;
  const surfaceLabel = surface === "reel" ? "Reels" : surface === "story" ? "Stories" : "the feed";
  const v = refusals[0];
  const lead = v.kind === "too_long" ? "Too long for"
    : v.kind === "too_short" ? "Too short for"
    : v.kind === "too_small" ? "Too small for"
    : v.kind === "too_large" ? "Too large for"
    : "Wrong shape for";
  return `${lead} ${surfaceLabel} (${v.message})`;
}

/**
 * True when SOME destination that can actually PUBLISH this media kind would accept it.
 * The upload gate's only job: refuse what nothing can take. Which destination a post
 * actually goes to is the composer's question, and it is the only place the app knows
 * the answer.
 *
 * This only asks a destination about a surface it can genuinely publish that media kind
 * to — `videoSurfaces(platform)` for video, `supportsImages(platform)` (+ the feed/story
 * surfaces a platform actually offers) for images. An earlier version of this function
 * iterated every platform/surface *recorded in media-limits.json* and accepted if any had
 * zero refusals. That is wrong: Threads has an image entry recorded for "feed" but NO
 * video publish path at all (`videoSurfaces("threads")` is `[]`, mirroring
 * `worker/clients.py`'s `PLATFORM_CAPS["threads"].video_surfaces`, an empty frozenset).
 * Asking `checkMedia("threads", "feed", videoAsset)` about a media kind Threads has no
 * entry for returns "no violations" — because `limitsFor` found nothing to check, not
 * because Threads would take the video. A 30-minute video was "accepted via threads/feed"
 * under that logic, which is nonsense: Threads cannot publish video at all. "No limit
 * recorded" and "this destination accepts anything" are different claims, and only the
 * per-platform capability flags (not presence in the JSON file) can tell them apart.
 *
 * WHY THIS NEVER REFUSES A VIDEO ON THIS INSTALL TODAY: TikTok can publish video
 * (`videoSurfaces("tiktok")` is `["feed"]`), but TikTok's limits are deliberately absent
 * from media-limits.json — they are per-creator and fetched at runtime from TikTok's
 * `creator_info` endpoint, not a static fact anyone can write down here (verified in an
 * earlier task; see reference.md). With TikTok in the platform list and its limits
 * unknown, `checkMedia("tiktok", "feed", asset)` can never report a refusal for ANY
 * duration — so this function can never prove that *nothing* accepts a given video, and
 * refuses none. That is the honest consequence of "only refuse what we can verify," not a
 * bug: it will look like a broken gate to the next reader until they read this comment.
 * An install without a TikTok channel configured would still see this function refuse an
 * over-long video normally — TikTok's mere presence in the *platform list* (not a
 * channel) is what keeps this permissive, since the gate has no way to know at upload
 * time which platforms the owner has actually connected.
 *
 * `data` defaults to the real media-limits.json for every production call. Tests pass a
 * synthetic `data` to exercise the refusal branch, which today's real data cannot reach.
 */
export function anyDestinationAccepts(asset: AssetLike, data: PlatformsData = REAL_PLATFORMS): boolean {
  const kind = asset.media_kind ?? "";
  for (const { value: platform } of PLATFORMS) {
    const surfaces: string[] =
      kind === "video" ? videoSurfaces(platform)
      : kind === "image" && supportsImages(platform) ? (supportsStory(platform) ? ["feed", "story"] : ["feed"])
      : [];
    for (const surface of surfaces) {
      const refusals = checkMedia(platform, surface, asset, data).filter((v) => v.severity === "refuse");
      if (refusals.length === 0) return true;
    }
  }
  return false;
}

File: worker/media_limits.py
"""Per-platform, per-surface media limits, read from dashboard/media-limits.json.

That file is shared with the dashboard on purpose. Hand-mirroring ~100 numbers across
two languages is the drift bug waiting to happen: the composer would say a send is fine,
the worker would refuse it, and nobody would learn why until a post failed. One file
makes that impossible rather than merely discouraged. Same reasoning the schema lives in
/migrations - it is a fact about the platform, owned by neither language.

The file lives under dashboard/, not the repo root, even though the worker (here) reads
it too. That is backwards-looking on purpose: Turbopack — the bundler that serves the
dashboard's "use client" components — refuses to bundle an import from outside its own
project root. Widening its root (turbopack.root) was tried and rejected on 2026-08-24:
it built, but pulled the 1.6GB data/ asset store into the build's file-trace manifest.
A checked-in generated copy under dashboard/ was rejected too — two files that must be
kept in sync is the exact drift this design exists to eliminate. So the worker reaches
INTO dashboard/ instead. One awkwardly-placed file beats either alternative.

ABSENT MEANS NOT ENFORCED. A limit we cannot verify is omitted, never guessed. See the
spec's Principle section: a wrong number that refuses valid media is worse than no check,
because the platform itself is always the backstop.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from fractions import Fraction
from functools import lru_cache
from pathlib import Path

# Only for PLATFORM_CAPS — see _media_is_conformed below. No cycle: clients.py imports
# .config/.discord_api/.graph_api/.telegram_api/.tiktok_api, none of which import this
# module, even though publisher.py (which DOES import this module) also imports clients.
from .clients import PLATFORM_CAPS

# worker/media_limits.py -> worker/ -> repo root -> dashboard/media-limits.json.
# See the module docstring for why this reaches into dashboard/ instead of the reverse.
RAW_PATH = Path(__file__).resolve().parent.parent / "dashboard" / "media-limits.json"


class MediaLimitsError(Exception):
    """The limits file is missing, malformed, or fails its schema check.

    Raised loudly and never swallowed: a broken config must not degrade into "allow
    everything", which would silently disable every media check in the app.
    """


@dataclass(frozen=True)
class Violation:
    kind: str      # too_short | too_long | too_small | too_large | wrong_aspect | wrong_format
    message: str   # a human sentence, safe to show in the UI
    severity: str  # "refuse" | "warn"


_NUMERIC = (
    "min_duration_ms", "max_duration_ms",
    "min_width", "min_height", "max_width", "max_height",
    "max_bytes",
)
_KNOWN = set(_NUMERIC) | {"min_aspect", "max_aspect", "formats", "note", "varies"}
_KNOWN_TOP_LEVEL = {"schema_version", "platforms", "_comment"}


@lru_cache(maxsize=1)
def load_limits() -> dict:
    try:
        raw = json.loads(RAW_PATH.read_text())
    except FileNotFoundError as exc:
        raise MediaLimitsError(f"media-limits.json not found at {RAW_PATH}") from exc
    except ValueError as exc:
        raise MediaLimitsError(f"media-limits.json is not valid JSON: {exc}") from exc

    if raw.get("schema_version") != 1:
        raise MediaLimitsError(f"unsupported schema_version {raw.get('schema_version')!r}")

    # The top level gets the same scrutiny as every entry below it. A typo here (e.g.
    # "paltforms") would otherwise load with zero exceptions and silently enforce
    # nothing — exactly the "malformed degrades into allow everything" outcome
    # Principle #2 forbids. An EMPTY platforms dict is legal (this file starts with
    # just Facebook and grows); only the wrong TYPE or a missing key is an error.
    unknown_top = set(raw) - _KNOWN_TOP_LEVEL
    if unknown_top:
        raise MediaLimitsError(f"media-limits.json has unrecognised top-level key(s): {sorted(unknown_top)}")
    if "platforms" not in raw:
        raise MediaLimitsError("media-limits.json has no 'platforms' key")
    if not isinstance(raw["platforms"], dict):
        raise MediaLimitsError("media-limits.json's 'platforms' must be an object")

    for platform, surfaces in raw.get("platforms", {}).items():
        for surface, kinds in surfaces.items():
            for kind, entry in kinds.items():
                where = f"{platform}.{surface}.{kind}"
                if not entry.get("note"):
                    raise MediaLimitsError(f"{where} has no note")
                unknown = set(entry) - _KNOWN
                if unknown:
                    raise MediaLimitsError(f"{where} has unrecognised field(s): {sorted(unknown)}")
                for field in ("min_aspect", "max_aspect"):
                    pair = entry.get(field)
                    if pair is not None and not (
                        isinstance(pair, list) and len(pair) == 2
                        and all(isinstance(n, int) and n > 0 for n in pair)
                    ):
                        raise MediaLimitsError(f"{where}.{field} must be [w, h] positive integers")
    return raw


def limits_for(platform: str, surface: str, media_kind: str) -> dict | None:
    """The recorded limits, or None when nothing is recorded — which means NOT ENFORCED."""
    return (
        load_limits()
        .get("platforms", {})
        .get(platform, {})
        .get(surface, {})
        .get(media_kind)
    )


def _get(asset, key):
    """Assets arrive as sqlite3.Row (no .get) or as plain dicts in tests."""
    if hasattr(asset, "keys"):
        return asset[key] if key in asset.keys() else None
    return asset.get(key)


def _media_is_conformed(platform: str, surface: str, media_kind: str | None, asset) -> bool:
    """Whether THIS asset's own width/height/byte_size describe the untouched original —
    or the conformed derivative (assets.publish_path) that has already replaced it for
    this destination, in which case checking the original's aspect ratio and byte size
    would be checking a file that is never actually sent.

    Mirrors publisher.py's _needs_conformed, plus the story/cover carve-out its caller
    _resolve_rel applies on top of it: _resolve_rel's story and cover branches ignore
    needs_conformed entirely and always prefer their OWN file (a Story's own 9:16 canvas,
    built by a separate pipeline — migration 0015 — or simply the untouched original, and
    a Reel's cover is always the untouched cover-conformed original) over the feed-shaped
    derivative at publish_path. So only feed/reel ever actually serve it. Duplicated here
    rather than imported from publisher.py to avoid a publisher <-> media_limits import
    cycle (publisher already imports this module) — keep the two in sync by hand if either
    changes; see publisher.py's _needs_conformed and _resolve_rel for the source of truth.
    """
    if surface in ("story", "cover"):
        return False
    publish_path = _get(asset, "publish_path")
    conform_mode = _get(asset, "conform_mode")
    has_conformed = bool(publish_path) or (conform_mode not in (None, "none"))
    if not has_conformed:
        return False
    caps = PLATFORM_CAPS.get(platform)
    if caps is None:
        return False
    if media_kind == "video" and surface == "feed" and not caps.feed_video_is_constrained:
        return False
    return caps.needs_conformed_media


def check(platform: str, surface: str, asset) -> list[Violation]:
    """Every limit this asset violates for this destination. Empty means publishable.

    Checked against the asset's OWN width/height/byte_size — except for aspect ratio and
    byte size, when this destination actually receives a conformed derivative (see
    _media_is_conformed above) rather than the untouched original: conformance guarantees
    the file that is actually sent already satisfies both, by construction
    (dashboard/lib/conform.ts), so checking the original's values there would be checking
    a file nobody ever publishes. An iPhone portrait photo (1179x2556) is the everyday
    case this matters for — its ORIGINAL fails instagram.feed.image's aspect range, but
    the conformed derivative that is actually published never does. Duration is
    unaffected either way: conforming never changes it. min/max width/height bounds are
    also unaffected: for video the stored width/height already describe whichever file
    was kept (the upload route overwrites them with the derivative's real dimensions when
    it converts one), and no image entry in media-limits.json defines width bounds today
    (Meta auto-scales an out-of-range width rather than refusing it — see
    instagram.feed.image's note).
    """
    kind = _get(asset, "media_kind")
    entry = limits_for(platform, surface, kind)
    if entry is None:
        return []
    # A limit that VARIES by account (Discord's, which depends on Nitro tier and server
    # boost) can never be enforced honestly — it warns instead, so the send still happens.
    severity = "warn" if entry.get("varies") else "refuse"
    conformed = _media_is_conformed(platform, surface, kind, asset)

    out: list[Violation] = []
    duration = _get(asset, "duration_ms")
    width, height = _get(asset, "width"), _get(asset, "height")
    size = _get(asset, "byte_size")

    # Every check below is guarded on the value being KNOWN. Unknown metadata must never
    # refuse: assets predating the video pipeline carry no duration at all.
    if duration is not None:
        lo, hi = entry.get("min_duration_ms"), entry.get("max_duration_ms")
        if lo is not None and duration < lo:
            out.append(Violation("too_short", f"shorter than {lo / 1000:g}s", severity))
        if hi is not None and duration > hi:
            out.append(Violation("too_long", f"longer than {hi / 1000:g}s", severity))

    # `is not None`, not truthiness: this file's whole purpose is that the two
    # languages cannot diverge, and the TypeScript side already guards on `!= null`.
    # A cosmetic `if width and height:` here would be a guard-style mismatch sitting
    # inside the one file that exists to prevent exactly that kind of drift.
    if width is not None and height is not None:
        mw, mh = entry.get("min_width"), entry.get("min_height")
        if (mw is not None and width < mw) or (mh is not None and height < mh):
            out.append(Violation(
                "too_small", f"smaller than {mw or '?'}x{mh or '?'}", severity
            ))
        xw, xh = entry.get("max_width"), entry.get("max_height")
        if (xw is not None and width > xw) or (xh is not None and height > xh):
            out.append(Violation(
                "too_large", f"larger than {xw or '?'}x{xh or '?'}", severity
            ))
        # Fraction, not float: a decimal comparison can exclude exactly 16:9 by rounding,
        # and 16:9 is a ratio Meta explicitly permits. Bounds are INCLUSIVE. Skipped when
        # `conformed`: the derivative that is actually sent is guaranteed in-range by
        # construction, so checking the original's ratio here would be checking a file
        # nobody publishes.
        if not conformed:
            ratio = Fraction(width, height)
            lo_a, hi_a = entry.get("min_aspect"), entry.get("max_aspect")
            if (lo_a is not None and ratio < Fraction(*lo_a)) or (
                hi_a is not None and ratio > Fraction(*hi_a)
            ):
                out.append(Violation("wrong_aspect", f"aspect ratio {width}x{height}", severity))

    # Skipped when `conformed`, same reasoning as the aspect check above: conform.ts's
    # encodeUnderLimit guarantees the sent derivative is under IG_MAX_BYTES by
    # construction.
    if (
        not conformed
        and size is not None
        and entry.get("max_bytes") is not None
        and size > entry["max_bytes"]
    ):
        out.append(Violation("too_large", f"larger than {entry['max_bytes']} bytes", severity))

    return out
